rubix-spi/src/main/java/com/qubole/rubix/spi/ClusterManager.java

```diff
@@ -91,13 +85,33 @@ protected String getCurrentNodeHostAddress()
       Collections.sort(nodes);
     }
 
-    currentNodeIndex = nodes.indexOf(getCurrentNodeHostname());
-    if (currentNodeIndex == -1) {
-      currentNodeIndex = nodes.indexOf(getCurrentNodeHostAddress());
+    // remove stale nodes from consistent hash ring
+    for (SimpleNode ringNode : consistentHashRing.getNodes()) {
+      if (!nodes.contains(ringNode.getKey()))
+      {
+        log.debug("Removing node: " + ringNode.getKey() + " from consistent hash ring");
+        consistentHashRing.remove(ringNode);
+      }
+    }
+
+    // add new nodes to consistent hash ring
+    for (String node : nodes) {
+      SimpleNode ringNode = SimpleNode.of(node);
+      if (!consistentHashRing.contains(ringNode)) {
+        log.debug("Adding node: " + ringNode.getKey() + " to consistent hash ring");
+        consistentHashRing.add(ringNode);
+      }
+    }
+
+    log.debug("Total nodes in consistent hash ring: " + consistentHashRing.getNodes());
+
```

Comment:
move currentNodeName setting under a if clause so that it is done only once. Over time JIT will remove that if clause.

Reply:
fixed.